### PR TITLE
st1 [2495][ADD] product_archive_sold_product

### PR DESCRIPTION
**Pull request by kanda999**
[2495](https://www.quartile.co/web?#id=2495&action=771&model=project.task&view_type=form&menu_id=505)

A cron task that will "Archive" products which have met certain conditions(qty_delivered > 0)

### CODE REVIEW

File: product_archive_sold_product/models/product_product.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                product_id,
+                sum(qty_delivered) AS sale_qty
+            FROM
+                sale_order_line
+            JOIN
+                product_product ON product_product.id = sale_order_line.product_id
+            WHERE
+                state in ('sale', 'done') AND active = True
+            GROUP BY
+                product_id
```
Comment (yostashiro) [inline product_archive_sold_product/models/product_product.py]:
```suggestion
            SELECT
                sol.product_id
            FROM
                sale_order_line sol
            JOIN
                product_product pp ON sold.product_id = pp.id
            WHERE
                sol.state in ('sale', 'done')
                AND sol.qty_delivered > 0
                AND pp.active = True
            GROUP BY
                sol.product_id
```

File: product_archive_sold_product/models/product_product.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        """
+        )
+
+        sale_dict = self.env.cr.dictfetchall()
+
+        sale_vals = {}
+        for dictionaly in sale_dict:
+            sale_vals[dictionaly["product_id"]] = dictionaly["sale_qty"]
```
Comment (yostashiro) [inline product_archive_sold_product/models/product_product.py]:
```suggestion
        prod_ids = [r[0] for r in self.env.cr.fetchall()]
```
`prod_ids = self.env.cr.fetchall()` might just work as well, but I don't feel comfortable with throwing a list of tuples in `browse()`.

File: product_archive_sold_product/models/product_product.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        for prod in self.search([("active", "=", True)]):
+            if sale_vals.get(prod.id, 0) > 0:
+                # if product has reordering rules then deactivate first
+                if prod.orderpoint_ids:
+                    prod.orderpoint_ids.write({"active": False})
+                prod.product_tmpl_id.write({"active": False})
```
Comment (yostashiro) [inline product_archive_sold_product/models/product_product.py]:
```suggestion
        for prod in self.browse(prod_ids):
            # if product has reordering rules then deactivate first
            if prod.orderpoint_ids:
                prod.orderpoint_ids.write({"active": False})
            prod.product_tmpl_id.write({"active": False})
```

**Review by yostashiro — CHANGES_REQUESTED** [review]

File: product_archive_sold_product/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                sale_order_line sol
+            JOIN
+                product_product pp ON sol.product_id = pp.id
+            JOIN
+                product_template pt ON pp.product_tmpl_id = pt.id
+            WHERE
+                sol.state in ('sale', 'done')
+                AND sol.qty_delivered > 0
+                AND pp.active = True
+                AND (pt.type = 'product' OR pt.type = 'consu')
```
Comment (yostashiro) [inline product_archive_sold_product/models/product_product.py]:
```suggestion
                AND pt.type in ('product', 'consu')
```

File: product_archive_sold_product/readme/DESCRIPTION.rst [hunk 5]
```diff
@@ -0,0 +1,7 @@
+This module does the following:
+
+* This module is used in env where have only one product record per product
+
+* A cron task that will "Archive" products which have met certain conditions.
+
+  - Target products to "Archive": The product has to be a stockable or consumable product and (qty_delivered > 0)
```
Comment (yostashiro) [inline product_archive_sold_product/readme/DESCRIPTION.rst]:
```suggestion
This module provides the function to archive "sold" products with cron.

Assumptions:

* Physical products (stockable/consumable) are created per piece with no exceptions. 
* Product variant feature is not enabled (i.e. there is just one variant per template).
```

File: product_archive_sold_product/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2022 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+{
+    "name": "Product Archive Sold Product",
+    "version": "15.0.1.0.0",
+    "author": "Quartile Limited",
+    "website": "https://www.quartile.co",
+    "category": "Extra Tools",
+    "license": "LGPL-3",
+    "depends": ["sale"],
```
Comment (yostashiro) [inline product_archive_sold_product/__manifest__.py]:
```suggestion
    "depends": ["sale_stock"],
```
Not strictly a technical dependency, but product type `product` comes from stock module.

**Review by yostashiro — CHANGES_REQUESTED** [review]
We might also need to consider the chunk size depending on the performance on data with large volume (in a similar manner as https://github.com/qrtl/hls-oca/pull/22).

**Comment by kanda999** [comment]
I've successfully archived 2691 products in the my dev environment.

**Review by yostashiro — APPROVED** [review]
👍

**Pull request merged**